source/common/router/router.cc

```diff
@@ -314,13 +335,20 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
 
   // A route entry matches for the request.
   route_entry_ = route_->routeEntry();
+  if (append_cluster_info) {
+    // If we're asked to append cluster info, this is our first opportunity. The cluster name will
+    // be appended to any local or upstream responses from this point.
+    modify_headers = [this](Http::HeaderMap& headers) {
```

Comment:
Do we want to do this here vs. some other filter per the doc? Also, don't we want config around this in terms of what header to set, etc.?

Reply:
Well, appending the response headers could certainly be done in a separate `StreamEncoderFilter`, but the router filter would still have to share the routing information with that separate filter somehow. I guess you could imagine tacking that info onto `StreamDebugInfo`... But I don't think I see a lot of value in that; it seems better to keep the router-related debugging stuff contained to the router filter.

As far as configurability of what response headers to set... I wasn't imagining we'd offer that kind of flexibility. I figure it's best to provide a fixed set of response headers with well-known meanings, so we don't run into situations where somebody emails envoy-users saying "I turned on the `x-foobar` header and it doesn't work," and nobody knows what they're talking about. But there certainly might be use cases I'm not picturing...

Reply:
The only reason I figured we might want it configurable is so that operators would have full control in case they have existing systems they want to integrate with. I don't feel very strongly about it so we can start with a fixed header if we think that's the way to go.

Reply:
Makes sense. Probably easy enough to make it configurable with a default.